evm/src/cpu/kernel/asm/curve/bn254/field_arithmetic/util.asm

```diff
@@ -145,253 +164,259 @@
+    %swap_mstore
+    // stack:                     x2, x3, x4, addr
+    PUSH 2
     MULFP254
-    // stack:                   2*x2, x3, x4, ptr
-    DUP4  
+    // stack:                   2*x2, x3, x4, addr
+    DUP4
     %add_const(2)
-    // stack:             ind2, 2*x2, x3, x4, ptr
-    %mstore_bn254_pairing
-    // stack:                         x3, x4, ptr
+    // stack:             addr2, 2*x2, x3, x4, addr
+    %swap_mstore
+    // stack:                         x3, x4, addr
     PUSH 2 
     MULFP254
-    // stack:                       2*x3, x4, ptr
-    DUP3  
+    // stack:                       2*x3, x4, addr
+    DUP3
     %add_const(3)
-    // stack:                 ind3, 2*x3, x4, ptr
-    %mstore_bn254_pairing
-    // stack:                             x4, ptr
-    PUSH 2  
+    // stack:                 addr3, 2*x3, x4, addr
+    %swap_mstore
+    // stack:                             x4, addr
+    PUSH 2
     MULFP254
-    // stack:                           2*x4, ptr
+    // stack:                           2*x4, addr
     SWAP1
-    // stack:                           ptr, 2*x4
+    // stack:                           addr, 2*x4
     %add_const(4)
-    // stack:                          ind4, 2*x4
-    %mstore_bn254_pairing
+    // stack:                          addr4, 2*x4
+    %swap_mstore
     // stack:
 %endmacro
 
 // cost: 6 stores + 6 pushes + 5 adds = 6*4 + 6*1 + 5*2 = 40
 %macro store_fp254_6(ptr)
     // stack:       x0, x1, x2, x3, x4, x5
     PUSH $ptr
-    // stack: ind0, x0, x1, x2, x3, x4, x5
-    %mstore_bn254_pairing
-    // stack:           x1, x2, x3, x4, x5
-    PUSH $ptr  
-    %add_const(1)
-    // stack:     ind1, x1, x2, x3, x4, x5
-    %mstore_bn254_pairing
-    // stack:               x2, x3, x4, x5
-    PUSH $ptr  
-    %add_const(2)
-    // stack:         ind2, x2, x3, x4, x5
-    %mstore_bn254_pairing
-    // stack:                   x3, x4, x5
-    PUSH $ptr  
-    %add_const(3)
-    // stack:             ind3, x3, x4, x5
-    %mstore_bn254_pairing
-    // stack:                       x4, x5
-    PUSH $ptr  
-    %add_const(4)
-    // stack:                 ind4, x4, x5
-    %mstore_bn254_pairing
-    // stack:                           x5
-    PUSH $ptr  
-    %add_const(5)
-    // stack:                     ind5, x5
-    %mstore_bn254_pairing
+    %create_bn254_pairing_address
+    // stack: addr, x0, x1, x2, x3, x4, x5
+    DUP1 DUP3
+    MSTORE_GENERAL
+    // stack: addr, x0, x1, x2, x3, x4, x5
+    %increment
+    DUP1 DUP4
+    MSTORE_GENERAL
+    // stack: addr, x0, x1, x2, x3, x4, x5
+    %increment
+    DUP1 DUP5
+    MSTORE_GENERAL
+    // stack: addr, x0, x1, x2, x3, x4, x5
+    %increment
+    DUP1 DUP6
+    MSTORE_GENERAL
+    // stack: addr, x0, x1, x2, x3, x4, x5
+    %increment
+    DUP1 DUP7
+    MSTORE_GENERAL
+    // stack: addr, x0, x1, x2, x3, x4, x5
+    %increment
+    %stack (addr, x0, x1, x2, x3, x4, x5) -> (x5, addr)
```

Comment:
Nit: Using the same technique as below in `store_fp254_6_sh` (DUP, SWAP, MSTORE) would help remove these POPs/SWAPs, right?

Reply:
I've actually replaced it with a call to the upper macro, as it's cheaper:
```asm
%macro store_fp254_6(ptr)
    // stack:       x0, x1, x2, x3, x4, x5
    PUSH $ptr
    %store_fp254_6
    // stack:
%endmacro
```